File: src/rate-limit.ts
import { getCoreProxy } from "./core-proxy-loader.js";
import type { RoutingProfile } from "./types.js";

export function isRateLimited(resp: Response): boolean {
  return resp.status === 429 || resp.headers.get("x-hub-rate-limited") === "1";
}

// Earliest epoch-ms the response says it'll be usable again (x-hub-retry-after-ms, else retry-after
// seconds). `now` defaults to Date.now() but is injectable for tests. Header extraction stays here;
// the arithmetic is CoreProxyJs.rateLimitResetMsJson (single source, mirrors RateLimit.java).
export function rateLimitResetMs(resp: Response, now: number = Date.now()): number {
  const args = {
    headers: {
      "x-hub-retry-after-ms": resp.headers.get("x-hub-retry-after-ms") || "",
      "retry-after": resp.headers.get("retry-after") || "",
    },
    now,
  };
  const core = getCoreProxy();
  return JSON.parse(core.rateLimitResetMsJson(JSON.stringify(args))) as number;
}

// Final response when every model in a chain is rate-limited. Delegates the native-shaped 429
// (status/headers/body) entirely to profile.nativeRateLimit, which owns its upstream's rate-limit
// header conventions and error format. The profile is the sole owner of the synthesized headers;
// this engine has no app-specific header names and overlays nothing on top of what the profile
// returns.
export async function rateLimitFinal(
  lastResp: Response | null,
  resetMs: number,
  profile: RoutingProfile
): Promise<Response> {
  const built = await profile.nativeRateLimit({ resetMs, upstream: lastResp });
  return new Response(built.body, { status: built.status, headers: new Headers(built.headers) });
}

File: java/routing/src/main/java/io/github/intisy/ai/shared/logic/RateLimit.java
package io.github.intisy.ai.shared.logic;

import io.github.intisy.ai.shared.routing.RateLimitInfo;
import io.github.intisy.ai.shared.routing.RoutingProfile;
import io.github.intisy.ai.shared.spi.http.HttpResponse;

import java.util.HashMap;
import java.util.Map;

/**
 * Rate-limit detection, reset-time computation, and native-shaped 429 synthesis. The engine has no
 * app-specific header names: {@link #rateLimitFinal} delegates entirely to
 * {@link RoutingProfile#nativeRateLimit} and returns its result verbatim. Operates on the SPI
 * {@link HttpResponse} (a plain {@code String} body passed straight through, no encoding/decoding).
 */
public final class RateLimit {

    private RateLimit() {
    }

    public static boolean isRateLimited(HttpResponse resp) {
        return resp.status == 429 || "1".equals(header(resp, "x-hub-rate-limited"));
    }

    /**
     * Earliest epoch-ms the response says it'll be usable again ({@code x-hub-retry-after-ms},
     * else {@code retry-after} seconds). {@code now} is caller-supplied (injectable for tests).
     */
    public static long rateLimitResetMs(HttpResponse resp, long now) {
        int xr = parseIntLenient(header(resp, "x-hub-retry-after-ms"));
        if (xr > 0) return now + xr;
        int ra = parseIntLenient(header(resp, "retry-after"));
        if (ra > 0) return now + ra * 1000L;
        return 0;
    }

    /**
     * Final response when every model in a chain is rate-limited. Delegates the native-shaped 429
     * (status/headers/body) entirely to {@code profile.nativeRateLimit}, which owns its upstream's
     * rate-limit header conventions and error format. The profile is the sole owner of the
     * synthesized headers; this engine overlays nothing on top of what the profile returns.
     */
    public static HttpResponse rateLimitFinal(HttpResponse lastResp, long resetMs, RoutingProfile profile) {
        RoutingProfile.Synth s = profile.nativeRateLimit.build(new RateLimitInfo(resetMs, lastResp));
        HttpResponse out = new HttpResponse();
        out.status = s.status;
        out.headers = new HashMap<>(s.headers);
        out.body = s.body;
        return out;
    }

    // Case-insensitive header lookup, mirroring the JS Headers API's case-insensitivity
    // regardless of how the underlying map's keys happen to be cased.
    private static String header(HttpResponse resp, String name) {
        if (resp == null || resp.headers == null) return null;
        String direct = resp.headers.get(name);
        if (direct != null) return direct;
        for (Map.Entry<String, String> e : resp.headers.entrySet()) {
            if (e.getKey() != null && e.getKey().equalsIgnoreCase(name)) return e.getValue();
        }
        return null;
    }

    // Parses an optional sign followed by a leading run of digits and ignores trailing garbage;
    // returns 0 (which the callers' `> 0` guard treats as absent) when no digits are present or the
    // header is absent.
    private static int parseIntLenient(String s) {
        if (s == null) return 0;
        s = s.trim();
        int i = 0, n = s.length();
        if (i < n && (s.charAt(i) == '+' || s.charAt(i) == '-')) i++;
        int start = i;
        while (i < n && Character.isDigit(s.charAt(i))) i++;
        if (i == start) return 0;
        try {
            return Integer.parseInt(s.substring(0, i));
        } catch (NumberFormatException e) {
            return 0;
        }
    }
}
